File: 3.无重复字符的最长子串.go
/*
 * @lc app=leetcode.cn id=3 lang=golang
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
func lengthOfLongestSubstring(s string) int {
	var slow, fast int
	var maxLength int

	visited := [128]int{}
	n := len(s)

	for ; fast < n; fast++ {
		visited[s[fast]]++
		for visited[s[fast]] > 1 {
			visited[s[slow]]--
			slow++
		}
		maxLength = max(maxLength, fast-slow+1)
	}

	return maxLength
}

// @lc code=end

File: 3.无重复字符的最长子串.py
#
# @lc app=leetcode.cn id=3 lang=python
#
# [3] 无重复字符的最长子串
#

# @lc code=start
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        visited = set()
        slow, max_length = 0, 0

        for fast, c in enumerate(s):
            while c in visited:
                visited.remove(s[slow])
                slow += 1
            visited.add(c)
            max_length = max(max_length, fast-slow+1)
        return max_length

# @lc code=end

